### ISSUE TITLE
Allow arbitrary frame for cartesian trajectory in cartesian_trajectory_controller

### ISSUE THREAD

**Issue by fmauch** [UniversalRobots/Universal_Robots_ROS_controllers_cartesian#11]
As far as I can see, the cartesian trajectory controller does not use the frame of the poses that are coming in.

In our [example test_move script](https://github.com/UniversalRobots/Universal_Robots_ROS_Driver/blob/master/ur_robot_driver/scripts/test_move) we do not set the `frame_id` explicitly and in the controller implementation I cannot find any mechanism to do a lookup. 

@stefanscherzinger am I missing something or is this functionality missing?

**Comment by stefanscherzinger** [comment]
@fmauch 
You are right. This functionality is not yet implemented. We currently assume the robot's base frame for control when processing sampled trajectory points. This is particularly the case for calls to [our IK solver](https://github.com/UniversalRobots/Universal_Robots_ROS_controllers_cartesian/blob/main/cartesian_trajectory_controller/include/cartesian_trajectory_controller/control_policies.hpp#L298) and when forwarding targets to the [UR IK solver](https://github.com/UniversalRobots/Universal_Robots_ROS_controllers_cartesian/blob/main/cartesian_trajectory_controller/include/cartesian_trajectory_controller/control_policies.hpp#L205). It's also the case [here](https://github.com/UniversalRobots/Universal_Robots_ROS_controllers_cartesian/blob/main/cartesian_trajectory_controller/include/cartesian_trajectory_controller/cartesian_trajectory_controller.hpp#L149).

The `cartesian_trajectory_interpolation`, however, is independently formulated in its own coordinate system that's [implicitly defined by the given poses](https://github.com/UniversalRobots/Universal_Robots_ROS_controllers_cartesian/blob/main/cartesian_trajectory_interpolation/include/cartesian_trajectory_interpolation/cartesian_state.h#L34) (without the loss of generality). All poses must have the same implicit reference frame, though. But that's assured by the [message definition](https://github.com/UniversalRobots/Universal_Robots_ROS_cartesian_control_msgs/blob/main/msg/CartesianTrajectory.msg)


I assume you have two common use cases in mind:

1. Re-use a previously taught Cartesian trajectory (e.g. some polishing motion) at a new - but *static* - spot in the robot's workspace.
2. Re-use and apply a previously taught Cartesian trajectory to some *moving* object.

Both `1.` and `2.` probably only make sense if the new frame is not directly part of the robotic chain, but belongs somewhere to the TF tree of the setup. I think we could achieve `1.` with a static transform that we need to look-up only once. That's probably okay. But since we sample control points during trajectory execution on the fly, `2.` will require to do the lookups online and possibly in real-time. Not sure at the moment how critical that is.

**Comment by fmauch** [comment]
What I have in mind is for example commanding a trajectory in a different reference frame than the one defined for the controller as in https://forum.universal-robots.com/t/about-the-cartesian-trajectory-controllers-and-movegroup/18606.

I am aware, that the message definition does provide a reference frame definition, but I couldn't find anything in the controller's code that is actually interpreting this.

**Comment by wafts95** [comment]
Hi!

So I am attempting to control the UR10 with the `forward_cartesian_traj_controller` and the UR ROS driver, and I have set up my tf tree to include some additional frames that are children of the tool0_controller (I have a setup with more than one "tool center point", attached is the tf tree generated by running `rosrun tf2_tools view_frames.py`). I am interested in commanding each tcp ("tcp_gripper1, tcp_gripper2) to go to the same pose in my environment (relative to my base frame), at separate instances.

I have therefore defined a desired `CartesianTrajectoryPoint` msg, which I intend to use in two `CartesianTrajectory` trajectory definitions, that differ in their `controlled_frame` attribute; one of them uses "tcp_gripper1" and the second uses "tcp_gripper2". Unfortunately, it seems that when I send each of these trajectories to the trajectory controller the arm drives the flange, and not the desired grippers, to the pose of interest. Note: on my teach pendant, I have made sure to have the active tcp set as the flange.

I am not sure if this is precisely the subject of this Issue, but I wanted to address it here before I raise another Issue.

Thank you for your time!

![Screenshot from 2022-03-25 23-57-21](https://user-images.githubusercontent.com/57106106/160212489-02cbe660-8a41-4de3-8e80-815c3e8b1a48.png)
